src/repr/src/row.proto

```diff
@@ -0,0 +1,135 @@
+        // on read). However, once a 1-byte id is used, it's gone forever, so
+        // we're convervative in handing them out.
+        //
+        // Of these, I'd guess Timestamp and UUID are probably the first ones
+        // we'd bless followed by Date and Time.
+        ProtoDate date = 16;
+        ProtoTime time = 17;
+        ProtoTimestamp timestamp = 18;
+        ProtoInterval interval = 19;
+        ProtoArray array = 20;
+        ProtoRow list = 21;
+        ProtoDict dict = 22;
+        ProtoNumeric numeric = 23;
+        bytes uuid = 24;
+    }
+}
+
+enum ProtoDatumOther {
+    // It's generally good practice to make id 0 (the default if the field is
+    // unset) in proto enums be an Unknown sentinel. This allows for
+    // distinguishing between unset and any of the enum variants.
+    //
+    // This enum is initially used only in a oneof, which means we can
+    // distinguish unset without this sentinel. But stick one in here anyway,
+    // in case this enum gets used somewhere else in the future.
+    Unknown = 0;
+    Null = 1;
+    False = 2;
+    True = 3;
+    JsonNull = 4;
+    Dummy = 5;
+}
+
+message ProtoDate {
+    // Year
+    int32 year = 1;
+    // Day-of-year (0..365)
+    uint32 ordinal = 2;
+}
+
+message ProtoTime {
+    // The number of seconds since midnight
+    uint32 secs = 1;
+    // Additional fractional seconds since midnight in nanosecond granularity.
+    // This can can exceed 1,000,000,000 in order to represent the leap second.
+    uint32 nanos = 2;
+}
+
+message ProtoTimestamp {
+    // Year
+    int32 year = 1;
+    // Day-of-year (0..365)
+    uint32 ordinal = 2;
+    // The number of seconds since midnight
+    uint32 secs = 3;
+    // Additional fractional seconds since midnight in nanosecond granularity.
+    // This can can exceed 1,000,000,000 in order to represent the leap second.
+    uint32 nanos = 4;
+    // If true, this timestamp is in UTC. If false, this timestamp is zoneless.
+    bool is_tz = 5;
+}
+
+message ProtoInterval {
+    // A possibly negative number of months.
+    int32 months = 1;
+    // A 128-bit nanosecond count, represented as a 64-bit lo and hi.
+    //
+    // TODO: Why do our in-mem intervals use this representation? I've seen
+    // month/day/nanos in the past becuase a day is not always the same number
+    // of nanos.
+    int64 duration_lo = 2;
+    int64 duration_hi = 3;
+}
+
+message ProtoArray {
+    // All array elements flattened into 1 dimension, encoded in row-major
+    // order.
+    ProtoRow elements = 1;
+    // TODO: This is a direct translation of our in-mem representation of array
+    // dimensions. Does anyone understand it well enough to document it?
+    repeated ProtoArrayDimension dims = 2;
+}
+
+message ProtoArrayDimension {
+    uint64 lower_bound = 1;
+    uint64 length = 2;
+}
+
+message ProtoDict {
+    map<string, ProtoDatum> elements = 1;
+}
+
+// TODO: These are difficult to define, but are passed directly from and into
```

Comment:
lmk if you'd like to provide some inline definitions for these fields

Reply:
Yes please! That would be lovely

Reply:
I copied over your docs from https://github.com/MaterializeInc/rust-dec/pull/59 but it raised an important question! The docs you link mention that this representation is machine-dependent ("The layout of an integer might be big-endian or little-endian, for example."). That might be a deal-breaker for its inclusion in this format. Curious what @benesch thinks but I'm pretty uncomfortable with big-endian machines being able to run on top of data persisted by little-endian machines (even if that's hypothetical).

Reply:
I'm not sure I understand the concern! Like, the reason we're not blatting that struct down directly to disk is because it contains a machine-dependent `u32`. But that's exactly why we're converting it to a Proto message and serializing *that* instead.

Reply:
Although now you have me scared. @sploiselle I'm not convinced the `to_raw_parts` and `from_raw_parts` are quite right. I believe those ought to return slices of u16 rather than u8 because of endianness. Does that make sense? I've got to run now but ping me if it doesn't.

Reply:
(Similarly I think this proto message ought to store repeated u16s rather than a byte array.)

Reply:
That works great for digits, exponents, bits, but e.g. the `lsu` field is still a black box to me. Are there any machine-dependant concerns with that one? I can read the docs a bit more (and will) but I'm already more comfortable hearing that you think what we're doing here will work specifically wrt machine-dependance.

Reply:
> The lsu field is one or more units in length (of type decNumberUnit, an unsigned integer), and contains the digits of the coefficient. Each unit represents one or more of the digits in the coefficient and has a binary value in the range 0 through 10n-1, where n is the number of digits in a unit, set by the compile-time definition DECDPUN. The size of a unit is the smallest of 1, 2, or 4 bytes which will contain the maximum value held in the unit.
>
> The units comprising the coefficient start with the least significant unit (lsu). Each unit except the most significant unit (msu) contains DECDPUN digits. The msu contains from 1 through DECDPUN digits, and must not be 0 unless digits is 1 (for the value zero). Leading zeros in the msu are never included in the digits count, except for the value zero.

We leave `DECDPUN` at its default (https://github.com/MaterializeInc/rust-dec/blob/6a8f1b7db583aa0aa05d98e148a55c3785b16849/decnumber-sys/decnumber/decNumber.h#L43) of 3, which means `decNumberUnit` is a `u16`. The `u16`s are always in order from least to most significant regardless of endianness. So the only thing we need to be careful of is the endianness of each `u16`. That's what I think we're presently mishandling by returning a `&[u8]` instead of a `&[u16]`, but the fix should be as easy as making that swap.

Reply:
Well you are obviously a much faster typer than me! :D

I _think_ roundtripping a [u16] will be fine? Pretty sure proto3 uses the packed encoding for repeated primitive fields, but I'll check. It'd be a bummer for this to be encoded with the non-packed encoding (as in, "might be faster to reparse a string" kind of bummer)

Reply:
Two wrinkles!

1) Proto doesn't support u16 fields, only u32 and u64, so we'll have to do a repeated u32 or something. Which is fine, they're varints anyway, so the representation on the wire would be the same. We'll just want to do defensive bounds checking on the decode side (because of the way proto varints work, probably better to map each u16 into a u32 instead of packing two u16s into each u32, though it would actually depend on the distribution of values). To be clear, all this bit would live in the encode/decode paths introduced in this PR, not proposing that it live in {to,from}_raw_parts

2) Looks like even though the spec says "In proto3, repeated fields of scalar numeric types use packed encoding by default." crate protobuf doesn't actually do that. Luckily, it does pack them if you manually specify [packed=true] on the field. One thing to think about here is that (I'm only just now realizing) the non-packed encoding is shorter for a repeated field of len 1, equal length for a repeated field of len 2, and longer for anything more than that. If e.g. we overwhelmingly expect this field to be a single u16 in the common case, we may want to use the non-packed encoding. (Fun fact: The spec actually requires that either encoding be accepted on the decode side for backward compatibility, so a truly clever encoder could switch behaviors based on the contents of what it's encoding.)

Reply:
So it occurs to me that we should discuss what we're optimizing the storage format for. The `decNumber` struct is optimized for performing arithmetic on the decimal, which is maybe the wrong tradeoff here. If we want to optimize instead for encoded size, we could use [BCD encoding](http://speleotrove.com/decimal/dnpack.html). The output of BCD encoding is an int representing the scale and a variable-length byte array representing the integer. That might be _much_ easier to handle here, and no endianness concerns at all.

Reply:
Looks like we haven't exposed bindings for those functions in `rust-dec` yet though, so someone would have to add those.

Reply:
Yeah, we're balancing all 3 of stability of the format/encoded size/encode+decode speed. I don't have the domain expertise to really have any sense for what gives us the best tradeoff of those. It's starting to sound like we may want to do some benchmarking of how fast raw_parts vs bcd are; at what point should we punt this into the same "we have an opportunity to finalize this later" bucket as time/date? (Not to pressure anything here, but I would love to get this merged next week.)

Reply:
My gut take is that BCD encoding is probably the right tradeoff. It's probably only an hour or two of work. If we want to skip benchmarking, defaulting to BCD encoding is probably my preference; it's honestly probably _less_ work to do BCD encoding than it is to fix the endianness issues with the current thing.

Reply:
I'll handle exporting a BCD-encoded LSU tomorrow.

Reply:
I’ve only scanned the bcd docs, so apologies if this is just obviously wrong, but it seems like we with bcd we have to externally roundtrip the special NaN, etc sort of numbers ya? Definitely doable, just making sure I’ve understood it

Reply:
Oh hrm, yeah, I think you're right. It's really too bad [`decimal128`](http://speleotrove.com/decimal/dncomps.html) isn't quite large enough, because that would be optimal. It uses a slightly more efficient encoding than BCD (called densely packed decimal), and preserves NaN and infinity.

Ugh. Would an enum like this:

```
enum DecimalSpecial {
    NOT_SPECIAL = 0,
    NAN = 1,
    INFINITY = 2,
}
```

be sufficient, @sploiselle, or are there distinctions between e.g. quiet NaNs and signaling NaNs that would get lost? If the latter, then probably scratch the BCD idea and go back to handling endianness properly with `decNumber`'s raw parts.

Reply:
> The special values are NaN (Not a Number), which may be quiet (propagates quietly through operations) or signaling (raises the Invalid operation condition when encountered), and ±infinity.

This makes it sound like there's only 4 special values we'd have to roundtrip: quiet NaN, signaling Nan, +inf, -inf. Am I reading that right?

Reply:
I'm checking into what the case is with the special values in a bit and will have something meaningful to contribute this afternoon. Sorry for the delay here.

Reply:
tl;dr we do not need to worry about preserving signalling/quiet NaNs.

In `rust-dec`, the only operations that set the `DECSNAN` (signalling NaN) bit are:
- Those that convert from fixed-precision decimals to arbitrary-scale decimals (known as "Numbers") in the library:
    - [`decimal32ToNumber`]
    - [`decimal128ToNumber`]
    - [`decimal64ToNumber`]

    This doesn't affect us because don't use any "fixed-precision" decimals
- Parsing [the string `"sNaN"`](https://github.com/MaterializeInc/rust-dec/blob/master/decnumber-sys/decnumber/decNumber.c#L548), which I fixed in MZ in MaterializeInc/materialize#9939

Now that I have that q answered, am moving onto looking into BCD encoding.

[`decimal32ToNumber`]: https://github.com/MaterializeInc/rust-dec/blob/master/decnumber-sys/decnumber/decimal32.c#L202
[`decimal64ToNumber`]: https://github.com/MaterializeInc/rust-dec/blob/master/decnumber-sys/decnumber/decimal64.c#L239
[`decimal128ToNumber`]: https://github.com/MaterializeInc/rust-dec/blob/master/decnumber-sys/decnumber/decimal128.c#L228